### PR TITLE
[Feature:RainbowGrades] Separate manual and GUI customization

### PR DESCRIPTION
**Pull request by ziesski**
This PR changes instructorUI, RainbowGrades Customization page.

Few major changes involves following:

1. Added `manual_customization.json` and `gui_customization.json` 
Now there are 4 files in course rainbowgrades directory.  `manual_customization.json`, `gui_customization.json`,  `customization.json` and `customization_no_comment.json`. On the website, instructor can either use the GUI to write customization or upload their own customization. Any edits made using GUI will be saved as `gui_customization.json` while uploaded customization will be saved as `manual_customization.json`. Once all the edits have been made, user will specify which customization to use and click Build. Build will copy specified customization to `customization.json` and inside this Build there will be process of stripping unnecessary comments which will produce `customization_no_comment.json` which will be used to build Rainbowgrades. 

2. AutoSaving of GUI customization (separation of save and build)
Previously, website had save change button which will write all the changes made to `customization.json`, and build with the changed data. This was inefficient because save change button does both writing to json, and building rainbowgrades, which discouraged users to make incremental changes, and time consuming (build can take upto few minutes) 
With the updated change, GUI seamlessly and simultaneously auto-writes to `gui_customization.json` each time user edits. User will push build button once they are ready to release rainbowgrades.

3. Fix silent fail to build
Currently, Save change button does not fully catch error, and even when it fails, it displays Successfully build. With this update, it checks if the `auto_rainbow_grades.py` has completed before displaying successfully build. 

4. Upload manual customization, and Download Manual/GUI customization
New Rainbow Customization

### CODE REVIEW

**Comment by zacharymnp** [comment]
A few initial comments --

- items inside save_status_container should be middle-aligned
- Submitty version should be put in correct location
- Customization box should be hidden when it is empty

image

- When I click the 'Download Manual Customization' button before manual_customization.json has been uploaded, I get the following error. However, when I click the 'Download GUI Customization' button before a GUI_customization.json has been generated, I am redirected to the 'Grade Reports' page.

image

There are also a few .json files that have been included in the top-level directory that should be removed.

**Review by oliiso — COMMENTED** [review]
A few notes:
- Adding a bit more space between the "Build" button and the "Status" in the save_status_container along with centering it would improve UI
- Not necessary, but putting a timer on the error message that is displayed when user clicks “download manual customization” and the download fails would be nice. Having the error fade/gone after a minute or so without the user needing to click the 'x' to take it away from the screen.
Image 7-2-24 at 11 00 AM

**Review by zacharymnp — CHANGES_REQUESTED** [review]
I'd recommend making the Log button white, because it's a nonessential button.
Also, the status message is pushed of the screen when the Log is viewed. The log message should probably go on the next line so that the status doesn't get moved.
Log Status

**Comment by oliiso** [comment]
This is because my error log is long, but it extended out of the designated frame. Is there a way to keep it within the designated log area?
Image 7-11-24 at 4 12 PM

File: package-lock.json
Comment (williamjallen) [inline package-lock.json:1]: Is this file needed?

File: site/app/controllers/admin/ReportController.php [hunk 1]
```diff
@@ -589,6 +589,35 @@ private function getLateStatusMessage(LateDayInfo $ldi) {
...
+                catch (\Exception $e) {
+                    die("error");
+                }
+            }
+            else {
+                die("error: no json_string provided");
+            }
+        }
+        else {
+            die("error: not an AJAX request");
```
Comment (williamjallen) [inline site/app/controllers/admin/ReportController.php:614]: It would be better to use json success and fail responses instead of just exiting immediately.

File: site/app/controllers/admin/ReportController.php [hunk 2]
```diff
@@ -589,6 +589,35 @@ private function getLateStatusMessage(LateDayInfo $ldi) {
...
+        }
+    }
+
+
+
+
```
Comment (williamjallen) [inline site/app/controllers/admin/ReportController.php:620]: Lots of extra whitespace here.

File: site/app/controllers/admin/ReportController.php [hunk 3]
```diff
@@ -652,95 +684,209 @@ public function generateCustomization() {
...
     }
 
+
+    #[Route("/courses/{_semester}/{_course}/reports/build_form", methods: ['POST'])]
+    public function executeBuildForm(): MultiResponse {
+        // Configure json to go into jobs queue
+        $job_json = (object) [];
+        $job_json->job = 'RunAutoRainbowGrades';
+        $job_json->semester = $this->core->getConfig()->getTerm();
+        $job_json->course = $this->core->getConfig()->getCourse();
```
Comment (williamjallen) [inline site/app/controllers/admin/ReportController.php:694]: Why not just use an array for this?  Something like this would be a lot cleaner:
```suggestion
        $job_json = [
            'job' => 'RunAutoRainbowGrades',
            'semester' => $this->core->getConfig()->getTerm(),
            'course' => $this->core->getConfig()->getCourse(),
        ];
```

File: site/app/controllers/admin/ReportController.php [hunk 4]
```diff
@@ -652,95 +684,209 @@ public function generateCustomization() {
...
+            '.json';
+
+        // Place in queue
+        file_put_contents($path, $job_json);
+
+        // Send a success response
+        return new MultiResponse(
+            JsonResponse::getSuccessResponse(['status' => 'success']),
+            null
+        );
```
Comment (williamjallen) [inline site/app/controllers/admin/ReportController.php:713]: Returning a MultiResponse isn't necessary here

File: site/app/controllers/admin/ReportController.php [hunk 5]
```diff
@@ -652,95 +684,209 @@ public function generateCustomization() {
         }
     }
 
+
+    #[Route("/courses/{_semester}/{_course}/reports/build_form", methods: ['POST'])]
+    public function executeBuildForm(): MultiResponse {
```
Comment (williamjallen) [inline site/app/controllers/admin/ReportController.php:689]: This function can just return a JsonResponse I think.

File: site/app/controllers/admin/ReportController.php [hunk 6]
```diff
@@ -652,95 +684,209 @@ public function generateCustomization() {
...
+        }
+
+        $rainbow_grades_dir = FileUtils::joinPaths($this->core->getConfig()->getCoursePath(), "rainbow_grades");
+        $destination_path = FileUtils::joinPaths($rainbow_grades_dir, 'manual_customization.json');
+
+        // this is changed from move_uploaded_file to copy because of permission issue (bits not carried over)
+        // copy is expensive, but we are OK because it is small file.
+        // setgid (sticky-bit) gets ignored and doesn't inherit the parent (rainbowgrades dir) permissions
+        // known issue: look https://www.php.net/manual/en/function.move-uploaded-file.php for more details
+        if (!copy($upload['tmp_name'], $destination_path)) {
```
Comment (williamjallen) [inline site/app/controllers/admin/ReportController.php:740]: Would it be better/easier to just read the file and then write it to the intended location?  That would come with added benefits of ensuring that it's a valid JSON file.

File: site/app/controllers/admin/ReportController.php [hunk 7]
```diff
@@ -652,95 +684,209 @@ public function generateCustomization() {
...
+        header("Content-Transfer-Encoding: Binary");
+        header("Content-disposition: attachment; filename=\"" . basename($file_path) . "\"");
+
+        readfile($file_path);
+        exit;
+        // Do not add return type, for example if you add :MultiResponse
+        // php-stan will start yelling that missing return statement
+        // if we add return statement, it will be written into the downloaded file as json
+        // then it will break the Make remove_json_comments, so no linting is OK here.
+    }
```
Comment (williamjallen) [inline site/app/controllers/admin/ReportController.php:785]: This would be a great place to use `DownloadResponse::getDownloadResponse()`.

File: site/app/controllers/admin/ReportController.php [hunk 8]
```diff
@@ -652,95 +684,209 @@ public function generateCustomization() {
...
-            null,
-            new RedirectResponse($redirect_url)
-        );
+        // Set headers and read the file
+        header('Content-Type: application/json');
+        header("Content-Transfer-Encoding: Binary");
+        header("Content-disposition: attachment; filename=\"" . basename($file_path) . "\"");
+
+        readfile($file_path);
+        exit;
```
Comment (williamjallen) [inline site/app/controllers/admin/ReportController.php:808]: Same here.

File: site/app/controllers/admin/ReportController.php [hunk 9]
```diff
@@ -652,95 +684,209 @@ public function generateCustomization() {
...
 
-    #[Route("/courses/{_semester}/{_course}/reports/rainbow_grades_status")]
+
+    #[Route('/courses/{_semester}/{_course}/reports/rainbow_grades_customization/manual_or_gui', methods: ['POST'])]
+    public function setRainbowGradeCustomization(): JsonResponse {
+
+        // Extract the value from $_POST
+        $selectedValue = $_POST['selected_value'] ?? null;
+
+        if (!isset($selectedValue) || trim($selectedValue) === '') {
```
Comment (williamjallen) [inline site/app/controllers/admin/ReportController.php:822]:
```suggestion
        if ($selectedValue === null || trim($selectedValue) === '') {
```

File: site/app/controllers/admin/ReportController.php [hunk 10]
```diff
@@ -652,95 +684,209 @@ public function generateCustomization() {
...
+            $msg = 'Invalid request: No selected value provided.';
+            return JsonResponse::getErrorResponse($msg);
```
Comment (williamjallen) [inline site/app/controllers/admin/ReportController.php:824]: The temp variable isn't needed here.

File: site/app/controllers/admin/ReportController.php [hunk 11]
```diff
@@ -652,95 +684,209 @@ public function generateCustomization() {
...
+        }
+
+        $rainbow_grades_dir = FileUtils::joinPaths($this->core->getConfig()->getCoursePath(), "rainbow_grades");
+        $customization_src = null;
```
Comment (williamjallen) [inline site/app/controllers/admin/ReportController.php:828]: This line isn't needed.  PHP uses function scoping, so the variable does not need to be defined outside the switch.

File: site/app/controllers/admin/ReportController.php [hunk 12]
```diff
@@ -652,95 +684,209 @@ public function generateCustomization() {
...
+
+        // Copy the source file to the destination
+        if (!copy($customization_src, $customization_dest)) {
+            $msg = 'File copy failed: Could not copy file.';
+            return JsonResponse::getErrorResponse($msg);
+        }
+
+        $msg = 'Rainbow Grades Customization set successfully';
+
+        return JsonResponse::getSuccessResponse(['status' => 'success', 'selected_value' => $selectedValue, 'message' => $msg ]);
```
Comment (williamjallen) [inline site/app/controllers/admin/ReportController.php:852]: It's generally better to put each element of an array on a separate line so the diff is cleaner if someone comes along later and changes a single element.  It's a good idea to use a trailing comma for the last element of the array for the same reason.
```suggestion
        return JsonResponse::getSuccessResponse([
            'status' => 'success',
            'selected_value' => $selectedValue,
            'message' => $msg,
        ]);
```

File: site/app/controllers/admin/ReportController.php [hunk 13]
```diff
@@ -652,95 +684,209 @@ public function generateCustomization() {
...
+    }
+
+
+
+    #[Route('/courses/{_semester}/{_course}/reports/rainbow_grades_status', methods: ['POST'])]
```
Comment (williamjallen) [inline site/app/controllers/admin/ReportController.php:857]: Lots of extra space here too.

File: site/app/controllers/admin/ReportController.php [hunk 14]
```diff
@@ -652,95 +684,209 @@ public function generateCustomization() {
...
     public function autoRainbowGradesStatus() {
-        // Create path to the file we expect to find in the jobs queue
```
Comment (williamjallen) [inline site/app/controllers/admin/ReportController.php:704]: Are the deleted comments wrong?  They could be more helpful, but I don't see a reason to delete them if they aren't wrong and aren't next to code you're refactoring.

File: site/app/controllers/admin/ReportController.php [hunk 15]
```diff
@@ -749,15 +895,14 @@ public function autoRainbowGradesStatus() {
         }
 
         $debug_contents = file_get_contents($debug_output_path);
+        $debug_contents = trim($debug_contents);
+        $was_successful = strrpos($debug_contents, 'Done') === (strlen($debug_contents) - strlen('Done'));
```
Comment (williamjallen) [inline site/app/controllers/admin/ReportController.php:899]: Would [`str_ends_with()`](https://www.php.net/manual/en/function.str-ends-with.php) be helpful here?

File: site/app/controllers/admin/ReportController.php [hunk 16]
```diff
@@ -749,15 +895,14 @@ public function autoRainbowGradesStatus() {
...
 
-        // If we finished the previous loops before max_wait_time hit 0 then the file successfully left the jobs queue
-        // implying that it finished
-        if ($max_wait_time && $failure_detected == false) {
-            $this->core->getOutput()->renderJsonSuccess($debug_contents);
+        if ($max_wait_time && $failure_detected === false && $was_successful) {
+            return JsonResponse::getSuccessResponse(['status' => 'success', 'data' => $debug_contents]);
```
Comment (williamjallen) [inline site/app/controllers/admin/ReportController.php:902]: This array can be wrapped as well.

File: site/app/models/RainbowCustomization.php [hunk 17]
```diff
@@ -597,38 +598,25 @@ public function processForm() {
...
-
-//        $this->has_error = "true";
-//        foreach($_POST as $field => $value){
-//            $this->error_messages[] = "$field: $value";
-//        }
-//        throw new ValidationException('Debug Rainbow Grades error', $this->error_messages);
     }
 
+
+
```
Comment (williamjallen) [inline site/app/models/RainbowCustomization.php:604]: This extra whitespace can be deleted

File: site/app/models/RainbowCustomization.php [hunk 18]
```diff
@@ -597,38 +598,25 @@ public function processForm() {
...
     }
+
+    public function doesManualCustomizationExist(): bool {
+        // using RCJSON will have issue because constructor will call loadFromJsonFile
+        // which will return null if gui_customization is not found.
+        // return $this->RCJSON->doesManualCustomizationExist();
+        $course_path = $this->core->getConfig()->getCoursePath();
+        $file_path = FileUtils::joinPaths($course_path, 'rainbow_grades', 'manual_customization.json');
+
+        return file_exists($file_path);
```
Comment (williamjallen) [inline site/app/models/RainbowCustomization.php:620]: The temp variable isn't really necessary here.

File: site/app/templates/admin/RainbowCustomization.twig [hunk 19]
```diff
@@ -16,17 +16,94 @@
...
+            <div id="build_buttons">
+                <button
+                    id="btn-build-customization"
+                    class="btn btn-primary key_to_click"
+                    tabindex="0"
+                    type="button"
+                    onclick="runBuild()"
+                >
+                    <i class="fas fa-magic-wand-sparkles" title="Download the file"></i>
+                    Build
```
Comment (williamjallen) [inline site/app/templates/admin/RainbowCustomization.twig:28]: This `title` attribute doesn't quite seem right

File: site/app/templates/admin/RainbowCustomization.twig [hunk 20]
```diff
@@ -16,17 +16,94 @@
...
+                                class="btn btn-default key_to_click"
+                                tabindex="0"
+                                type="button"
+                                title="Upload a Rainbow Grades customization.json directly to Submitty"
+                            >
+                                <i class="fas fa-upload" title="Download the file"></i>
+                                Upload Manual Customization
+                            </button>
+                            <button id="btn-download-customization" class="btn btn-default key_to_click" tabindex="0" type="button"
+                                    onclick="window.location.href='{{ manual_customization_download_url }}'">
```
Comment (williamjallen) [inline site/app/templates/admin/RainbowCustomization.twig:64]: It might be better to put each attribute on a separate line like you did above.

File: site/app/templates/admin/RainbowCustomization.twig [hunk 21]
```diff
@@ -16,17 +16,94 @@
...
+                                </div>
+                            </div>
+                        </form>
+                    </fieldset>
+
+            </div>
+
+
+
             <div id="display" class="customization_item">
```
Comment (williamjallen) [inline site/app/templates/admin/RainbowCustomization.twig:107]: Lots of extra space here...

File: site/app/templates/admin/RainbowCustomization.twig [hunk 22]
```diff
@@ -16,17 +16,94 @@
...
+                                </div>
+                                <div>
+                                    <input type="radio" id="gui_customization" name="customization" value="gui" {{ manual_customization_exists ? '' : 'checked' }}>
+                                    <label for="gui_customization">GUI Customization</label>
+                                </div>
+                            </div>
+                        </form>
+                    </fieldset>
+
+            </div>
```
Comment (williamjallen) [inline site/app/templates/admin/RainbowCustomization.twig:103]: Are you missing a closing tag?

File: site/app/templates/admin/Report.twig
Comment (williamjallen) [inline site/app/templates/admin/Report.twig:1]: The changes to this file should be reverted since they aren't related to the rest of the PR.

File: site/app/templates/admin/RainbowCustomization.twig [hunk 23]
```diff
@@ -243,7 +320,7 @@
                                                        onblur="ClampPoints(this);"
                                                        onkeyup="DetectMaxOverride(this);">
                                                 <div class="gradeable-li-right-of-input">
-                                                        <div class="gradeable-li-info">
+                                                    <div class="gradeable-li-info">
```
Comment (williamjallen) [inline site/app/templates/admin/RainbowCustomization.twig:323]: Unrelated to the rest of the PR.

File: site/public/js/rainbow-customization.js [hunk 24]
```diff
@@ -471,18 +475,98 @@ function showLogButton(responseData) {
...
+                },
+                error: function (jqXHR, textStatus, errorThrown) {
+                    console.log('AJAX error:', jqXHR, textStatus, errorThrown);
+                    $('#save_status').html('An error occurred while making the request');
+                },
+            });
+        })
+        .catch((error) => {
+            console.error('Caught error:', error);
+            $('#save_status').html(`An error occurred: ${error}`);
```
Comment (williamjallen) [inline site/public/js/rainbow-customization.js:555]: Using `.html()` opens the door to XSS attacks and should be avoided.  Use `.text()` or similar instead.

File: site/public/js/rainbow-customization.js [hunk 25]
```diff
@@ -471,18 +475,98 @@ function showLogButton(responseData) {
...
+                        $('#save_status').html('Generating rainbow grades, please wait...');
+                        checkBuildStatus();
+                    }
+                    else {
+                        $('#save_status').html('An error occurred while building');
+                    }
+                },
+                error: function (jqXHR, textStatus, errorThrown) {
+                    console.log('AJAX error:', jqXHR, textStatus, errorThrown);
+                    $('#save_status').html('An error occurred while making the request');
```
Comment (williamjallen) [inline site/public/js/rainbow-customization.js:549]: Same here.

File: site/public/js/rainbow-customization.js [hunk 26]
```diff
@@ -493,56 +577,86 @@ function checkAutoRGStatus() {
...
+            console.error(`Error status: ${textStatus}`);
+            console.error(`Error thrown: ${errorThrown}`);
+            console.error(`Server response: ${jqXHR.status} ${jqXHR.statusText}`);
+        },
+    });
 }
 
+$(document).ready(() => {
+    $("input[name='customization']").change(() => {
+        $('#save_status').html('Switched customization, need to rebuild');
```
Comment (williamjallen) [inline site/public/js/rainbow-customization.js:656]:
```suggestion
        $('#save_status').text('Switched customization, need to rebuild');
```

**Review by williamjallen — CHANGES_REQUESTED** [review]

**Comment by williamjallen** [comment]
> This is because my error log is long, but it extended out of the designated frame. Is there a way to keep it within the designated log area?

This should be all you need:
```css
overflow-wrap: anywhere;
white-space: pre-wrap;
```

File: site/app/templates/admin/RainbowCustomization.twig (hunk 22 above)
Comment (ziesski) [reply site/app/templates/admin/RainbowCustomization.twig:114]: I think it was confusing because of the indentation, fixed

File: site/app/controllers/admin/ReportController.php (hunk 15 above)
Comment (ziesski) [reply site/app/controllers/admin/ReportController.php:925]: This is a good point, I was unaware of it, looks like it is php8 thing.

**Review by zacharymnp — CHANGES_REQUESTED** [review]
I'm still unhappy with the way the Log currently looks. 
https://github.com/user-attachments/assets/693b42d1-c924-480a-ad95-d0a9d9553207

Other than that and the issue with Download Manual Customization, I am satisfied.

File: site/app/controllers/admin/ReportController.php (hunk 6 above)
Comment (ziesski) [reply site/app/controllers/admin/ReportController.php:740]: I initially liked this idea, but after testing out with existing instructor's customization, `if ($decoded_content === null && json_last_error() !== JSON_ERROR_NONE) {` will fail with syntax error. I think we should leave it like this. Below is what I was trying to do. 
스크린샷 2024-07-17 오후 1 35 13


      $decoded_content = json_decode($json_content);
        if ($decoded_content === null && json_last_error() !== JSON_ERROR_NONE) {
            $error_code = json_last_error();
            $error_message = json_last_error_msg();
            error_log("JSON decode error [$error_code]: $error_message");
            $msg = "Upload failed: Invalid JSON format ($error_message)";
            $this->core->addErrorMessage($msg);
            return JsonResponse::getErrorResponse($msg);
        }

**Comment by oliiso** [comment]
After successfully running submitty_install on this branch, I have been getting a server error when I click on the “Grade Reports” page. On other branches this error does not appear.
Image 7-18-24 at 11 52 AM

**Review by bmcutler — CHANGES_REQUESTED** [review]
Let's change the order of these buttons:
Screenshot 2024-07-19 at 2 08 48 PM

on the left, "download GUI"   
space in the middle
on the right, "upload manual" and "download manual"

------------

If there is no manual customization, the download manual button should be disabled/greyed out.

If there is no GUI customization (first time on the page), the download gui button should be disabled/greyed out.

---------

Preliminary testing it seemed to work for me....
I will look at code next.

**Comment by bmcutler** [comment]
minor:
The log button seems squashed too close here.
This is a changeScreenshot 2024-07-19 at 2 20 51 PM
Also, I don't think it needs to be blue.
(it should succeed so looking at the log is not necessary - except for power users/debugging)

**Comment by oliiso** [comment]
When the build fails, the log errors still extend outside of the designated area. 
Image 7-21-24 at 11 25 AM
It would look cleaner to update it to stay inside the grey box.

**Review by zacharymnp — CHANGES_REQUESTED** [review]
Text inside the log is still extending outside the designated area. Also, the log button is appearing before there is a log, with no change when it is clicked.
Log
Log Before Build

The issue with sections has not been fixed.
Sections

File: site/app/controllers/admin/ReportController.php [hunk 27]
```diff
@@ -589,6 +589,41 @@ private function getLateStatusMessage(LateDayInfo $ldi) {
...
+    public function writetocustomization(): JsonResponse {
+        if (!empty($_SERVER['HTTP_X_REQUESTED_WITH']) && strtolower($_SERVER['HTTP_X_REQUESTED_WITH']) == 'xmlhttprequest') {
+            // Build a new model, pull in defaults for the course
+            $customization = new RainbowCustomization($this->core);
+            $customization->buildCustomization();
+
+            if (isset($_POST["json_string"])) {
+                try {
+                    $customization->processForm();
+                    return JsonResponse::getSuccessResponse(['status' => 'success']);
```
Comment (IDzyre) [inline site/app/controllers/admin/ReportController.php:604]: The JSON Success response already defaults to having a success status, so you don't need to specify a 'status' here.

**Review by zacharymnp — CHANGES_REQUESTED** [review]
Errors are thrown when saveChanges() is called.

Other than that, I am satisfied. The migration appears to be successful and the UI looks good.

Before Migration:
Before Migration
After Migration:
After Migration

**Review by zacharymnp — APPROVED** [review]

File: site/phpstan-baseline.neon [hunk 28]
```diff
@@ -1464,6 +1464,16 @@ parameters:
...
+		-
+			message: "#^Method app\\\\controllers\\\\admin\\\\ReportController\\:\\:downloadGUIRainbowConfig\\(\\) has no return type specified\\.$#"
+			count: 1
+			path: app/controllers/admin/ReportController.php
+
+		-
+			message: "#^Method app\\\\controllers\\\\admin\\\\ReportController\\:\\:downloadRainbowConfig\\(\\) has no return type specified\\.$#"
+			count: 1
+			path: app/controllers/admin/ReportController.php
+
```
Comment (ziesski) [inline site/phpstan-baseline.neon:1476]:
```suggestion
```

File: site/app/controllers/admin/ReportController.php [hunk 29]
```diff
@@ -601,6 +601,41 @@ private function getLateStatusMessage(LateDayInfo $ldi) {
         }
     }
 
+
+
+    #[Route("/courses/{_semester}/{_course}/reports/rainbow_grades_customization_save", methods: ["POST"])]
```
Comment (williamjallen) [inline site/app/controllers/admin/ReportController.php:606]: There are a couple extra blank lines here...

File: site/app/controllers/admin/ReportController.php [hunk 30]
```diff
@@ -601,6 +601,41 @@ private function getLateStatusMessage(LateDayInfo $ldi) {
...
+    public function writetocustomization(): JsonResponse {
+        if (!empty($_SERVER['HTTP_X_REQUESTED_WITH']) && strtolower($_SERVER['HTTP_X_REQUESTED_WITH']) == 'xmlhttprequest') {
```
Comment (williamjallen) [inline site/app/controllers/admin/ReportController.php:608]: Why do you need to access the `$_SERVER` global here?

File: site/app/controllers/admin/ReportController.php [hunk 31]
```diff
@@ -601,6 +601,41 @@ private function getLateStatusMessage(LateDayInfo $ldi) {
...
+            else {
+                $msg = 'No JSON string provided';
+                $this->core->addErrorMessage($msg);
+                return JsonResponse::getErrorResponse($msg);
+            }
+        }
+        else {
+            $msg = 'Not an AJAX request';
+            $this->core->addErrorMessage($msg);
+            return JsonResponse::getErrorResponse($msg);
```
Comment (williamjallen) [inline site/app/controllers/admin/ReportController.php:633]: There's a lot of duplication here.  Would it be better to move everything but the message text definition to the bottom?

File: site/app/controllers/admin/ReportController.php [hunk 32]
```diff
@@ -666,62 +704,174 @@ public function generateCustomization() {
...
+
+        // Encode
+        $job_json = json_encode($job_json, JSON_PRETTY_PRINT);
+
+        // Create path to new jobs queue json
+        $path = '/var/local/submitty/daemon_job_queue/auto_rainbow_' .
+            $this->core->getConfig()->getTerm() .
+            '_' .
+            $this->core->getConfig()->getCourse() .
+            '.json';
```
Comment (williamjallen) [inline site/app/controllers/admin/ReportController.php:725]: It would be better to use `FileUtils::joinPaths()` for this.

File: site/app/controllers/admin/ReportController.php [hunk 33]
```diff
@@ -666,62 +704,174 @@ public function generateCustomization() {
...
+        if (!copy($upload['tmp_name'], $destination_path)) {
+            $msg = 'Upload failed: Could not copy file';
+            $this->core->addErrorMessage($msg);
+            return JsonResponse::getErrorResponse($msg);
+        }
+
+        $manual_customization_exists =  file_exists($destination_path);
+
+        $msg = 'Rainbow Grades Customization uploaded';
+        $this->core->addSuccessMessage($msg);
```
Comment (williamjallen) [inline site/app/controllers/admin/ReportController.php:765]: This temporary variable isn't necessary.

File: site/app/controllers/admin/ReportController.php [hunk 34]
```diff
@@ -666,62 +704,174 @@ public function generateCustomization() {
...
+
+    #[Route('/courses/{_semester}/{_course}/reports/rainbow_grades_status', methods: ['POST'])]
     public function autoRainbowGradesStatus() {
         // Create path to the file we expect to find in the jobs queue
         $jobs_file = '/var/local/submitty/daemon_job_queue/auto_rainbow_' .
             $this->core->getConfig()->getTerm() .
             '_' .
             $this->core->getConfig()->getCourse() .
             '.json';
-
```
Comment (williamjallen) [inline site/app/controllers/admin/ReportController.php:724]: This whitespace change doesn't appear to be related to the rest of the PR and should be reverted.

File: site/app/templates/admin/RainbowCustomization.twig [hunk 35]
```diff
@@ -16,17 +16,102 @@
...
+            <span id="status_label" style="padding-right: 4px;">Status: </span>
             <div id="save_status" data-testid="save-status">
                 No changes to save
             </div>
             <div id="save_status_buttons">
-                <input id="save_status_button" class="key_to_click" tabindex="0" type="button" value="Save Changes" onclick="ajaxUpdateJSON();" data-testid="save-status-button"/>
-                <input id="show_log_button" type="button" value="Log" />
+                <button style="margin-left: 60px; padding: 5px 12px;" class="btn btn-default" id="show_log_button" type="button">
+                    Log
+                </button>
```
Comment (williamjallen) [inline site/app/templates/admin/RainbowCustomization.twig:38]: Why does this button need special padding?

File: site/app/templates/admin/RainbowCustomization.twig [hunk 36]
```diff
@@ -16,17 +16,102 @@
...
             </div>
             <div id="save_status_log"></div>
         </div>
         <div id="list_wrapper">
+
+            <div id="upload_and_download" class="customization_item">
```
Comment (williamjallen) [inline site/app/templates/admin/RainbowCustomization.twig:44]: This whitespace doesn't denote a logical break, and is thus unnecessary.

File: site/app/templates/admin/RainbowCustomization.twig [hunk 37]
```diff
@@ -16,17 +16,102 @@
...
+                        </form>
+                    </div>
+                </div>
+            </div>
+
+            {% if manual_customization_exists %}
+            <div id="ask_which_customization" class="customization_item" style="display:block;">
+            {% else %}
+            <div id="ask_which_customization" class="customization_item" style="display:none;">
+            {% endif %}
```
Comment (williamjallen) [inline site/app/templates/admin/RainbowCustomization.twig:95]: This could be simplified by using a ternary expression.

File: site/app/templates/admin/RainbowCustomization.twig [hunk 38]
```diff
@@ -16,17 +16,102 @@
...
+                <fieldset>
+                    <legend><h2>Customization</h2></legend>
+                    <input type="hidden" name="csrf_token" value="{{ csrfToken }}" />
+                    <form id="customizationForm">
+                        <div id="customization-options" style="display:block;">
+                            <p>It appears that a manual customization file has been provided. Please specify the preferred file to be utilized for rendering Rainbow Grades:</p>
```
Comment (williamjallen) [inline site/app/templates/admin/RainbowCustomization.twig:101]: Is there a more concise way to phrase this?

File: site/app/templates/admin/RainbowCustomization.twig [hunk 39]
```diff
@@ -16,17 +16,102 @@
...
+            {% if manual_customization_exists %}
+            <div id="ask_which_customization" class="customization_item" style="display:block;">
+            {% else %}
+            <div id="ask_which_customization" class="customization_item" style="display:none;">
+            {% endif %}
+                <fieldset>
+                    <legend><h2>Customization</h2></legend>
+                    <input type="hidden" name="csrf_token" value="{{ csrfToken }}" />
+                    <form id="customizationForm">
+                        <div id="customization-options" style="display:block;">
```
Comment (williamjallen) [inline site/app/templates/admin/RainbowCustomization.twig:100]: Is the `display:block` necessary here?

File: site/app/templates/admin/Report.twig
Comment (williamjallen) [inline site/app/templates/admin/Report.twig:1]: The changes to this file are unrelated to the rest of the PR and should be reverted.

**Review by williamjallen — CHANGES_REQUESTED** [review]
This is certainly much farther along than the last time I reviewed it.  Most of my requested changes are minor stylistic issues.  After they're resolved I'll approve the code.  I can't really speak for the functionality since I have little experience with Rainbow Grades.

File: site/app/controllers/admin/ReportController.php (hunk 32 above)
Comment (ziesski) [reply site/app/controllers/admin/ReportController.php:723]: Well, this is more of `.` separator for string concatenation, but sure

File: site/app/templates/admin/RainbowCustomization.twig (hunk 35 above)
Comment (ziesski) [reply site/app/templates/admin/RainbowCustomization.twig:38]: Because I wanted the button size to exactly match Build button, or my center align will kinda downshift the characters.

File: site/app/controllers/admin/ReportController.php (hunk 30 above)
Comment (ziesski) [reply site/app/controllers/admin/ReportController.php:607]: needed during development stage, no longer need it good point.

File: site/app/controllers/admin/ReportController.php (hunk 31 above)
Comment (ziesski) [reply site/app/controllers/admin/ReportController.php]: good point

File: site/app/templates/admin/Report.twig [hunk 40]
```diff
@@ -71,7 +71,7 @@
                     <form action="{{ customization_upload_url }}" method="POST" enctype="multipart/form-data">
                         <input type="hidden" name="csrf_token" value="{{csrfToken}}">
                         <input type="file" id="config-upload" name="config_upload" style="display:none">
-                    
+
```
Comment (ziesski) [inline site/app/templates/admin/Report.twig:74]:
```suggestion
                                 
```

File: site/app/templates/admin/Report.twig [hunk 41]
```diff
@@ -71,7 +71,7 @@
...
```
Comment (ziesski) [inline site/app/templates/admin/Report.twig:74]:
```suggestion
                             
```

File: site/app/templates/admin/Report.twig [hunk 42]
```diff
@@ -71,7 +71,7 @@
...
```
Comment (ziesski) [inline site/app/templates/admin/Report.twig:74]:
```suggestion
                     
```

File: site/app/controllers/admin/ReportController.php (hunk 32 above)
Comment (williamjallen) [reply site/app/controllers/admin/ReportController.php]: Oh, my bad on that.  I thought there were slashes between the components.

**Review by williamjallen — APPROVED** [review]
Code looks reasonable to me.

**Review by IDzyre — APPROVED** [review]

**Review by bmcutler — CHANGES_REQUESTED** [review]
I tried to test this.
Before installing i built a simple rainbow grades via GUI,
THen I installed the PR and every time I hid the build button I get:

Screenshot 2024-07-29 at 1 26 06 PM

I don't have a "log" button.  I can't see any of the warning/error messages.

When I build from the terminal, there don't seem to be any errors.

**Comment by bmcutler** [comment]
* display of the log throws off the vertical alignment of the whole box.

* layout of the buttons (download gui, upload manual, download manual) wraps weird

Screenshot 2024-08-01 at 3 04 21 PM

File: sbin/auto_rainbow_grades.py [hunk 43]
```diff
@@ -208,6 +194,7 @@ def log_message(message):
 # Recursively update permissions for all files in the rainbow_grades directory
 print('Updating permissions', flush=True)
 cmd_output = os.popen('chmod -R --silent o-rwx ' + rg_course_path).read()
+# cmd_output = os.popen('find ' + rg_course_path + ' -type f -name "*customization*.json" -exec chmod g+rw {} \\;').read()
```
Comment (bmcutler) [inline sbin/auto_rainbow_grades.py:197]: remove unused code

**Review by bmcutler — CHANGES_REQUESTED** [review]

File: sbin/auto_rainbow_grades.py (hunk 43 above)
Comment (ziesski) [inline sbin/auto_rainbow_grades.py:197]:
```suggestion
```
